ldms/src/ldmsd/ldmsd_request.c

```diff
@@ -766,6 +767,29 @@ static int string2attr_list(char *str, struct attr_value_list **__av_list,
 	return rc;
 }
 
+int validate_ldmsd_req(ldmsd_req_ctxt_t reqc, ldmsd_req_hdr_t rh)
+{
+	ldmsd_req_attr_t attr;
+	unsigned int off;
+	/* Validate length of request header and request. */
+	if (rh->rec_len < sizeof(*rh))
+		return 0;
```

Comment:
```
if (rh->rec_len == sizeof(*rh))
    return 1;
```
Then add code to ldmsd_attr_first() to return NULL in this case.

Reply:
Also,
```
else if (rh->rec_len < sizeof($rh) + sizeof(attr->discrim))
    return 0;
```
At that point you know that the message includes attr->discrim and it can be safely checked.

Reply:
Wait, why wouldn't we just check sizeof(*attr)? Are we really allowing partial attr structs?? If we don't check the full attr size here, then we have to do anther size check later before we access attr->attr_len. That seems like madness. I hope we don't allow that in the protocol.

Reply:
Yes, the last 'attr' is just the discriminator which is the 'end of attribute list marker' No it's not madness. I believe you use NFS and RPC works exactly this way.

Reply:
Could you point me to where it works like that?

In any event, a quirk like that in the protocol needs to be clearly documented, and LDMS is certainly missing that.

But my original point stands that we have to check that the attr->attr_len fits in the overall message size before we access attr->attr_len. Knowing that attr->discrim fits is not sufficient to know that the rest of the attr can be safely checked.

Reply:
> Could you point me to where it works like that?
> 

RFC 8687 is the latest spec. linux/net/sunrpc/xprtrdma/rpc_rdma.h is the header file.

> In any event, a quirk like that in the protocol needs to be clearly documented, and LDMS is certainly missing that.
> 

There's a comment in the ldmsd_request.h file that documents this, perhaps not well enough.

> But my original point stands that we have to check that the attr->attr_len fits in the overall message size before we access attr->attr_len. Knowing that attr->discrim fits is not sufficient to know that the rest of the attr can be safely checked.

... and it's a good point.